### ISSUE TITLE
Some question about Hardsigmoid

### ISSUE THREAD

**Issue by MARD1NO** [pytorch/pytorch#49922]
First I found a redundant code in hardsigmoid implementation 

In this code 76 line 
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.cpp#L76
I found 
```
Tensor result;
```
The `result` is not be used. 

Otherwise I wonder how to implement the `hardsigmoid` inplace version 
Here is a test code
```
import torch 
import numpy as np 

x = np.array([3.0000, 3.00001]).astype(np.float32)

torch_x = torch.Tensor(x)
torch_x.requires_grad = True
torch_hsigmoid = torch.nn.Hardsigmoid(inplace=True)

torch_x2 = torch_x + 0 # Do no operations
torch_out = torch_hsigmoid(torch_x2)
print("Torch out is: ", torch_out)

torch_out = torch.mean(torch_out)
print("Torch out is: ", torch_out)
torch_out.backward()
print("Torch grad is: ", torch_x.grad)
```
In the boundary value `3`, it still have grad. But in value `3.00001`, it has no grad. However, these two numbers activated by `hardsigmoid`, their output value is the same. It seems like I cannot infer the grad according to the output value ? 

The another question is, I discover that `relu`, `hardtanh` have no grad in boundary value except the `hardsigmoid` operator. Is this a bug? 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @mruberry @jbschlosser

**Comment by mruberry** [comment]
> In this code 76 line
> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.cpp#L76
> I found
> 
> ```
> Tensor result;
> ```
> 
> The `result` is not be used.

Thank you for pointing this out. I think you're correct and we would accept a PR fixing this.


@albanD, what do you think of the gradient issues?

**Comment by MARD1NO** [comment]
I also found the same question in the operator `torch.nn.Hardswish` . In boundary value `-3` it still have grad. It also seems like I cannot infer the grad according to the output value, which is used in `inplace` version.

**Comment by soulitzer** [comment]
In general, for these non-differentiable "boundary values" (and if the space is locally convex), we can pick a [subgradient](https://en.wikipedia.org/wiki/Subderivative) to designate as the gradient. 1/12 is indeed a valid subgradient at 3, so this behavior should be OK.

However we'd generally also like to pick the subgradient with the minimum norm for the theoretical guarantees it provides and for consistency reasons. And as you also noted, since we set grad to 0 for the boundary value of relu, it might make sense to fix this as well for the sake of consistency.

**Comment by MARD1NO** [comment]
Thanks for your reply ~, now i have no problems about the `subgredient`. 

The last question is: 
![图片](https://user-images.githubusercontent.com/42901638/103335066-c97b5480-4aae-11eb-8dde-98686958dc73.png)
According to the formula of `hardswish`, it seems like we cannot infer the gradient with the output value(in other words, when `hardswish`'s parameter `inplace` is `True`). Because in forward computation, it involves `x^2` and `x`.

**Comment by albanD** [comment]
> According to the formula of hardswish, it seems like we cannot infer the gradient with the output value

The input value is actually cloned and saved in this case automatically by the autograd because as you mention this wouldn't be possible otherwise.

**Comment by MARD1NO** [comment]
Thanks a lot !
